README.md
# summer-conf
--------------------------------------------------

### What is the summer-conf ?
==================================================
Summer-conf is a program developing by C++ and 
can read configure easily.    
The confiure style likes that:
    # main section
    [main]
    name = aleda
    age  = 13
    # other section
    [other]
    codename = aleda
    nickname = aleda
    array    = 1, 2, 3, 4
and you can read the element through these steps:
    1. summerconf::Config conf = summerconf::Config('test.conf');    
    2. int num_value = conf["main"]["age"]->int_value();    
    3. std::string string_value = conf["main"]["name"]->string_value();    
    4. std::vector<std::string> string_arr = conf["other"]["array"]->string_value_list();    
and you should note these:
    1. if section dose not find, secion_name that value was "__INTER_MARK__" will return;
    2. if value dose not find, value that was "__NOT_FOUND__" will return;
more details, you should see test/summerconf_test.cpp.    

### How to use it ?
==================================================
    1. make clean
    2. make ar
    or you will just:
    make output
    more details, see makefile;

### Any qustion?
==================================================
    Please contact with Aleda[aleda.cn | aledalee@foxmail].    


NEWS
2015-05-03 02:01
------------------------------------------------------------------
    (1) prelinary test passed;    
2015-05-19 16:05    
    (1) uploading to github;    


test/summerconf_test.cpp
/* Copyright (C) 
 * 2015 - Aleda(aledalee@foxmail.com | aleda.cn)
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * 
 */
/**
 * @file summerconf_test.cpp
 * @brief summerconf_test
 * @author Aleda(aledalee@foxmail.com | aleda.cn)
 * @version 1.0.0
 * @date 2015-04-26
 */

#include <iostream>
#include <vector>

#include "config.h"
#include "section.h"

int main(int argc, char* argv[])
{
    summerconf::Config conf("./test.conf");
    //summerconf::Config conf("../../summer-se/conf/crawler.conf");
    conf.read();
    std::cout << "conf[main][name] = " << conf["main"]["name"]->string_value() << std::endl;
    int age = conf["main"]["age"]->int_value();
    std::cout << "conf[main][age] = " << age << std::endl;
    summerconf::Section section_crawler = conf["crawler"];
    std::cout << "conf[crawler].section_name = " << section_crawler.get_section_name() << std::endl;
    std::cout << "conf[crawler][author] = " << conf["crawler"]["author"]->string_value() << std::endl;
    std::cout << "conf[main][nickname] = " << conf["crawler"]["nickname"]->string_value() << std::endl;
    return 0;
    // TEST: conf_file
    // ***************************************************************//
    std::string conf_file = conf.get_conf_file();
    std::cout << "conf_file = " << conf_file << std::endl;

    summerconf::Section _section = conf["main"];
    std::cout << "_section_name = " << _section.get_section_name() << std::endl;
    std::map<std::string, summerconf::Value*> _item_map = _section.get_item_map();
    std::map<std::string, summerconf::Value*>::iterator _it_map;
    for (_it_map = _item_map.begin(); _it_map != _item_map.end(); ++_it_map) {
        std::cout << "key = " << _it_map->first << std::endl;
        std::cout << "value(string) = " << _it_map->second->string_value() << std::endl;
    }
    const summerconf::Value* _val = _section["name"];
    std::cout << "_val = " << _val->string_value() << std::endl;

    std::cout << "_section[\"name\"] = " << _section["name"]->string_value() << std::endl;
    std::cout << "conf[main][name] = " << conf["main"]["name"]->string_value() << std::endl;


    // ***************************************************************//

    // TEST  : Value
    // ***************************************************************//
    // STATUS: PASS
    summerconf::Value val = summerconf::Value("123");
    int num_val = val.int_value();
    std::cout << "num_val   = " << num_val << std::endl;

    std::string string_value = summerconf::Value("aleda").string_value();
    std::cout << "string_value = " << string_value << std::endl;

    std::vector<std::string> string_list = summerconf::Value("1,2,3,4").string_value_list();
    std::vector<std::string>::iterator it;
    for (it = string_list.begin(); it != string_list.end(); ++it) {
        std::cout << "__data = " << *it << std::endl;
    }
    // ***************************************************************//

    // TEST  : Section
    // ***************************************************************//
    // STATUS: PASS
    std::vector<summerconf::Section*> section_list = conf.get_section_list();
    std::vector<summerconf::Section*>::iterator it2;
    for (it2 = section_list.begin(); it2 != section_list.end(); ++it2) {
        std::string section_name = (*it2)->get_section_name();
        std::cout << "section_name = " << section_name << std::endl;
        std::map<std::string, summerconf::Value*> item_map = (*it2)->get_item_map();
        std::map<std::string, summerconf::Value*>::iterator it3;
        for (it3 = item_map.begin(); it3 != item_map.end(); ++it3) {
            std::cout << "key = " << it3->first << std::endl;
            std::cout << "value = " << it3->second->string_value() << std::endl;
        }
    }
    // ***************************************************************//
    return 0;
}



src/value.h
/* Copyright (C) 
* 2015 - Aleda(aledalee@foxmail.com | aleda.cn)
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
* 
*/
/**
* @file value.h
* @brief value
* @author Aleda(aledalee@foxmail.com | aleda.cn)
* @version 1.0.0
* @date 2015-04-26
*/

#ifndef VALUE_H_
#define VALUE_H_

#include <string>
#include <vector>

namespace summerconf {

class Value {
public:
    Value(const std::string& data): data(data) {}
    ~Value() {};
public:
    int int_value() const;
    std::string string_value() const;
    std::vector<std::string> string_value_list() const;
public:
    const std::string& get_data() const {return data;}
private:
    std::string data;
};

} // namespace summerconf

#endif



src/config.h
/* Copyright (C) 
* 2015 - Aleda(aledalee@foxmail.com | aleda.cn)
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
* 
*/
/**
* @file config.h
* @brief config
* @author Aleda(aledalee@foxmail.com | aleda.cn)
* @version 1.0.0
* @date 2015-04-26
*/

#ifndef CONFIG_H_
#define CONFIG_H_

#include <string>
#include <vector>

#include "section.h"
#include "value.h"

#define DISALLOW_COPY_AND_ASSIGN(TypeName) \
    TypeName(const TypeName&); \
    TypeName& operator=(const TypeName&)

namespace summerconf {

class Config {
public:
    explicit Config(const std::string& conf_file): conf_file(conf_file) {}
    ~Config();
public:
    bool read();
    Section& get(const std::string& section_name);
    Section& operator[](const std::string& section_name);
public:
    const std::vector<Section* >& get_section_list() const {return section_list;}
    const std::string& get_conf_file() const {return conf_file;};
    const std::vector<std::string>& get_data() const {return data;}
protected:
    bool parse();
private:
    bool judge(const std::string& section, 
               const std::string& key,
               const std::string& value);
private:
    std::string conf_file;
    std::vector<std::string> data;
    std::vector<Section* > section_list;
    class Iterator {
    public:
        Iterator(const std::vector<std::string>& data);
        ~Iterator() {};
    public:
        const std::string& get_section() const;
    public:
        bool loop();
        std::string get_key();
        std::string get_value();
        Iterator& operator++();
    private:
        bool parse();
        bool judge();
        int erase(std::string& str_src, const std::string& std_d);
    private:
        std::vector<std::string> data;
        std::string section;
        std::vector<std::string>::iterator index;
        std::vector<std::string>::iterator size;
    };
    DISALLOW_COPY_AND_ASSIGN(Config);
};

} // namespace summerconf

#endif



src/section.h
/* Copyright (C) 
* 2015 - Aleda(aledalee@foxmail.com | aleda.cn)
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
* 
*/
/**
* @file section.h
* @brief section
* @author Aleda(aledalee@foxmail.com | aleda.cn)
* @version 1.0.0
* @date 2015-04-26
*/

#ifndef SECTION_H_
#define SECTION_H_

#include <string>
#include <map>

#include "value.h"

namespace summerconf {

class Section {
public:
    Section() {}
    Section(const std::string& section_name, 
            std::map<std::string, Value*> item_map): 
    section_name(section_name),
    item_map(item_map) {}
    ~Section() {}
public:
    //std::map<std::string, Value*>& get();
    const Value* operator[](const std::string& key);
public:
    const std::string& get_section_name() const {return section_name;}
    const std::map<std::string, Value*>& get_item_map() const{return item_map;}
private:
    std::string section_name;
    std::map<std::string, Value*> item_map;
};

} // namesapce summerconf

#endif



src/value.cpp
/* Copyright (C) 
* 2015 - Aleda(aledalee@foxmail.com | aleda.cn)
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
* 
*/
/**
* @file value.cpp
* @brief value
* @author Aleda(aledalee@foxmail.com | aleda.cn)
* @version 1.0.0
* @date 2015-04-26
*/

#include <string.h>
#include <string>
#include <stdlib.h>

#include "value.h"

namespace summerconf {

int Value::int_value() const
{
    return atoi(data.c_str());
}

std::string Value::string_value() const
{
    return data;
}

std::vector<std::string> Value::string_value_list() const
{
    std::string data_temp = data;
    std::vector<std::string> ans_list;
    ans_list.clear();
    std::size_t pos;
    // space split
    std::string delim = ",";
    while (std::string::npos != (pos = data_temp.find(delim))) {
        ans_list.push_back(data_temp.substr(0, pos));
        data_temp.erase(0, pos + delim.length());
    }
    ans_list.push_back(data_temp);
    return ans_list;
}

} // namespace summerconf



src/section.cpp
/* Copyright (C) 
* 2015 - Aleda(aledalee@foxmail.com | aleda.cn)
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
* 
*/
/**
* @file section.cpp
* @brief section
* @author Aleda(aledalee@foxmail.com | aleda.cn)
* @version 1.0.0
* @date 2015-04-26
*/

#include <iostream>
#include <string>
#include <map>

#include "value.h"
#include "section.h"

namespace summerconf {

const Value* Section::operator[](const std::string& key)
{
    if (item_map.find(key) != item_map.end()) {
        return item_map[key];
    } else {
        return item_map["__NOT_FOUND__"];
    }
}

} // namespace summerconf



src/config.cpp
/* Copyright (C) 
 * 2015 - Aleda(aledalee@foxmail.com | aleda.cn)
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * 
 */

/**
 * @file config.cpp
 * @brief config
 * @author Aleda(aledalee@foxmail.com | aleda.cn)
 * @version 1.0.0
 * @date 2015-04-26
 */

#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <fstream>

#include "config.h"
#include "value.h"

namespace summerconf {

Config::~Config()
{
    //TODO 
    //erase the section_list;
    std::vector<Section*>::iterator it;
    for (it = section_list.begin(); it != section_list.end(); ++it) {
        std::map<std::string, Value*> _item_map = (**it).get_item_map();
        std::map<std::string, Value*>::iterator it_map;
        for (it_map = _item_map.begin(); it_map != _item_map.end(); ++it_map) {
            if (it_map->second != NULL) {  
                delete it_map->second;
                it_map->second = NULL;
            }
        }
        if (*it != NULL) {
            delete *it;
            *it = NULL;
        }
    }
}

bool Config::read()
{
    std::string line;
    std::ifstream ifs(conf_file.c_str());
    if (ifs) {
        while (getline(ifs, line)) {
            data.push_back(line);
        }
        ifs.close();
        return parse();
    } else {
        return false;
    }
}

bool Config::parse() {
    Iterator it(data);
    std::string old_section = it.get_section();
    std::map<std::string, Value*> item_map;
    item_map.clear();
    std::string section;
    while (it.loop()) {
        section = it.get_section();
        std::string key     = it.get_key();
        std::string value   = it.get_value();
        if (old_section.compare(section) != 0 &&
            old_section.compare("") != 0) {
            item_map["__NOT_FOUND__"] = new Value("__NOT_FOUND__");
            Section* obj_section = new Section(old_section, item_map);
            section_list.push_back(obj_section);
            item_map.clear();
        }
        if (judge(section, key, value)) {
            // yinyong is ok????
            // I am not sure!
            item_map[key] = new Value(value);
        }
        old_section = section;
        ++it;
    }
    if (old_section.compare("") != 0) {
        Section* obj_section = new Section(section, item_map);
        section_list.push_back(obj_section);
    }
    // Added strategy: if not have section, return NULL
    item_map.clear();
    item_map["__NOT_FOUND__"] = new Value("__NOT_FOUND__");
    section = "__INTER_MARK__";
    Section* obj_section = new Section(section, item_map);
    section_list.push_back(obj_section);
    return true;
}

bool Config::judge(const std::string& section,
                   const std::string& key,
                   const std::string& value)
{
    if (section.compare("") == 0) {
        return false;
    } else {
        if (key.compare("") == 0) {
            return false;
        }
    }
    return true;
}

Section& Config::operator[](const std::string& key)
{
    std::vector<Section*>::iterator it;
    std::vector<Section*>::iterator it_not_found;
    for (it = section_list.begin(); it != section_list.end(); ++it) {
        if ((*it)->get_section_name().compare(key) == 0) {
            return **it;
        }
        if ((*it)->get_section_name().compare("__INTER_MARK__") == 0) {
            it_not_found = it;
        }
    }
    return **it_not_found;
}

Config::Iterator::Iterator(const std::vector<std::string>& data)
{
    this->data = data;
    index = this->data.begin();
    size  = this->data.end();
    section = "";
}

std::string Config::Iterator::get_key()
{
    if (!Config::Iterator::judge()) {
        return std::string("");
    }
    //TODO
    //THIS SHOULD REPLACE space & tab;
    std::string normal_str = (*index);
    erase(normal_str, " ");
    erase(normal_str, "\t");
    std::size_t pos = normal_str.find("=");
    if (std::string::npos != pos) {
        return normal_str.substr(0, pos);
    } else {
        return std::string("");
    }
}

std::string Config::Iterator::get_value()
{
    if (!Config::Iterator::judge()) {
        return std::string("");
    }
    //THIS SHOULD REPLACE space & tab;
    std::string normal_str = (*index);
    erase(normal_str, " ");
    erase(normal_str, "\t");
    std::size_t pos = normal_str.find("=");
    if (std::string::npos != pos) {
        return normal_str.substr(pos + 1);
    } else {
        return std::string("");
    }

}

bool Config::Iterator::parse()
{
    if (index == size) {
        return false;
    }
    std::size_t start = (*index).find_first_not_of(" \t");
    if ('[' == (*index).at(start)) {
        std::size_t end = (*index).find("]");
        section = (*index).substr(start + 1, end - 1);
        return true;
    }
    return false;
}

bool Config::Iterator::judge()
{
    if ((*index).length() <= 0 || (*index).at(0) == '#') {
        return false;
    } else {
        return true;
    }
}

int Config::Iterator::erase(std::string& str_src, const std::string& str_d)
{
    int count = 0;
    std::size_t pos = 0;
    std::size_t str_d_size = str_d.size();
    while (std::string::npos != (pos = str_src.find(str_d, 0))) {
        str_src.erase(pos, str_d_size);
        ++count;
    }
    return count;
}

Config::Iterator& Config::Iterator::operator++()
{
    if (index != size) {
        ++index;
        parse();
    }
    return *this;
}

const std::string& Config::Iterator::get_section() const
{
    return section;
}

bool Config::Iterator::loop()
{
    if (index == size) {
        return false;
    }
    std::string normal_data = *index;
    //erase(normal_data, " ");
    //erase(normal_data, "\t");

    std::size_t start = normal_data.find_first_not_of(" \t");
    if ('[' == normal_data.at(start)) {
        std::size_t end = normal_data.find("]");
        section = normal_data.substr(start + 1, end - 1);
        index++;
    }
    return index != size;
}

} // namespace summerconf


